File: oneapp/onespace/basemap.py
"""Where a map gets its ground.

Every map surface in the product — the engine's map view type, and any space
that draws its own — needs the same two answers: what to paint under the
records, and who to credit for it. Neither is a workspace's choice. Which tile
store a bench points at is decided by whoever runs the bench, it is the same
for every workspace on the instance, and an air-gapped install has to be able
to change it in one place rather than in every screen that happens to draw a
map. So it is site config, read once, and handed to the app on the boot payload
beside the accent.

Three settings, in the order they win:

  * ``oneapp_map_style`` — a whole MapLibre style, as a URL. The full-control
    answer: a self-hosted vector style, a paid provider, an offline bundle.
    Given one, nothing else here applies.

  * ``oneapp_map_tiles`` — a raster tile template, ``{z}/{x}/{y}``. The common
    answer, because a raster source is a URL rather than a style document and
    because a tile that fails to load is one grey square rather than a map that
    never loads at all. A style URL that 404s takes the whole map with it.

  * ``oneapp_map_attribution`` — who to credit, when the default is not right.

With none of them set the default is OpenFreeMap's Positron, a vector style
over OSM data drawn quietly: grey ground, thin roads, low-contrast labels, so a
route line and a moving vehicle are the loudest things on the screen. That is
the point of choosing it over ``tile.openstreetmap.org``, whose own rendering is
built to show OSM's data rather than to sit under someone else's, and whose tile
usage policy asks applications not to use it as a basemap anyway.

**And below the instance, a workspace.** Which tile store to talk to is the
operator's decision; whether places are *named* and how much of the world is
drawn under the records are not — they are about the screens this customer
looks at all day, and a network diagram and a delivery round want different
answers. So `OneSpace Map Settings` is a tenant single holding three: a style
by name, whether to draw labels, and how much detail. It can only narrow what
the instance allows — it picks between the styles in `STYLES` or turns the
ground off entirely, and an instance that has named its own style outright
keeps it. Because the styles are vector, the last two apply without a reload:
`restyle` in `screen/basemap.js` walks the layers already on the map.

The browser fetches tiles directly, so the tile host sees the reader's IP and
viewport. That is a disclosure, and it is why the default is named in the
subprocessor clause any space that draws a map contributes, why it is one
config key to move, and why a map still draws with no tile host at all: a flat
ground and the records on it. That is not a degraded mode to apologise for —
it is what a schematic looks like, and it means this works offline, in a test,
and on a bench nobody has pointed at a tile store.
"""

import frappe
from frappe import _

#: The basemaps a workspace may choose between, as MapLibre style URLs.
#:
#: **Vector, and that is the whole point rather than a detail.** A raster tile
#: is a picture somebody else has already drawn: the only thing a reader can
#: change about it is what goes on top. A vector style is JSON the browser
#: executes, so which layers draw, whether places are named and how much of the
#: world is under the records are all things a workspace can decide at runtime
#: — see `frontend/src/modules/onespace/lib/screen/basemap.js`.
#:
#: OpenFreeMap, because it needs no key and is self-hostable. The default was
#: CARTO's raster Positron and it had quietly stopped being usable: their
#: keyless endpoint now returns every tile stamped "API KEY REQUIRED", which
#: nothing in this product could have noticed — a watermark is a valid PNG.
#:
#: **Canvas is ours**, and is the only one of the four that could be. The other
#: three are documents somebody else wrote and serves, so the most a workspace
#: can do to them is turn layers off. Canvas is a file in this repository —
#: `scripts/gen_basemap.py` writes it — so its palette is a decision we make,
#: and it is the one that can eventually follow the workspace's own accent. Its
#: tiles still come from OpenFreeMap, so it adds nobody to disclose.
STYLES = {
	"Canvas": "/assets/oneapp/basemaps/canvas.json",
	"Positron": "https://tiles.openfreemap.org/styles/positron",
	"Bright": "https://tiles.openfreemap.org/styles/bright",
	"Liberty": "https://tiles.openfreemap.org/styles/liberty",
}

#: Quiet grey, thin roads, low-contrast labels — so a route line and a moving
#: vehicle are the loudest things on the screen.
DEFAULT_STYLE = STYLES["Positron"]

#: Who to credit when nothing else does — which is only the raster path.
#:
#: **A style document credits itself.** Every style here names a TileJSON, and
#: that document carries the full linked attribution the licence actually wants:
#: OpenFreeMap, OpenMapTiles, and OpenStreetMap, each as a link. MapLibre reads
#: it and draws it, so sending our own alongside it stacked two credits into one
#: line — "© OpenStreetMap contributors © OpenFreeMap | OpenFreeMap ©
#: OpenMapTiles Data from OpenStreetMap" — saying the same thing twice, once
#: less completely.
#:
#: A raster template is a URL and nothing else: there is no document to read a
#: credit out of, so one is needed. This is the minimum true of any OSM-derived
#: store; an operator pointing at their own should name themselves, which is
#: what ``oneapp_map_attribution`` is for.
DEFAULT_ATTRIBUTION = "© OpenStreetMap contributors"

#: What a workspace gets before anybody opens the picker. Quiet rather than
#: Full: every screen that draws a map draws *records* on it, and buildings and
#: points of interest are competing with them for the same pixels.
DEFAULTS = {"pick": "Follow the instance", "labels": True, "detail": "Quiet"}

#: Who the default sends a request to, for the subprocessor clauses. A space
#: that draws a map names these; see `onemobility/legal.py`.
DEFAULT_HOSTS = ("tiles.openfreemap.org",)


def chosen() -> dict:
	"""What this workspace has decided about its map.

	`pick` is the *name* of a style and `style` on the boot payload is the URL
	it resolves to — two different things, and calling them both "style" put a
	word where a URL belonged and drew nothing.

	Read defensively. This is on the boot payload, which the control site
	renders too, and `OneSpace Map Settings` is a tenant doctype — so on the
	control plane the table is simply not there, and a map still has to draw.
	"""
	found = dict(DEFAULTS)
	if not frappe.db.exists("DocType", "OneSpace Map Settings"):
		return found
	row = frappe.get_cached_doc("OneSpace Map Settings")
	found["pick"] = (row.get("map_style") or DEFAULTS["pick"]).strip()
	found["labels"] = bool(row.get("map_labels"))
	found["detail"] = (row.get("map_detail") or DEFAULTS["detail"]).strip()
	return found


def boot() -> dict:
	"""The basemap, for the boot payload. See `www/one.py`."""
	style = (frappe.conf.get("oneapp_map_style") or "").strip()
	tiles = (frappe.conf.get("oneapp_map_tiles") or "").strip()
	dark = (frappe.conf.get("oneapp_map_tiles_dark") or "").strip()
	prefer = chosen()

	# The instance's own answer, before this workspace had a say. Sent alongside
	# the resolved one because the picker resolves the *next* style itself: this
	# payload was written when the page loaded, so a reader going back to "Follow
	# the instance" would otherwise be handed whatever they had picked before.
	instance = style or (DEFAULT_STYLE if not tiles else "")

	# An explicit empty string is how an instance says "no third-party tiles":
	# `frappe.conf` cannot hold a false that is distinguishable from unset, so
	# the opt-out is its own key rather than an empty value for another.
	#
	# A workspace that picks Plain says the same thing for itself, which is the
	# air-gapped instance's setting made available to one customer on a shared
	# one. Either way nothing is fetched.
	if frappe.conf.get("oneapp_map_plain") or prefer["pick"] == "Plain":
		return {"style": "", "tiles": "", "dark": "", "attribution": "", **prefer,
		        "styles": dict(STYLES), "instance": instance, "plain": True}

	# The workspace's own choice wins over the instance's default and loses to
	# an instance that has named a style outright: an operator who has pointed
	# a bench at their own tile store has done so for a reason, and a customer
	# picking "Bright" should not send them back off it.
	picked = STYLES.get(prefer["pick"], "")
	# An operator's own credit always wins and is always drawn — they may be
	# serving tiles nobody else knows the provenance of. Otherwise: nothing on
	# the style path, because the style says it better than we would.
	credit = (frappe.conf.get("oneapp_map_attribution") or "").strip()
	resolved = style or picked or (DEFAULT_STYLE if not tiles else "")
	return {
		"style": resolved,
		"tiles": tiles,
		"dark": dark or tiles,
		"attribution": credit or ("" if resolved else DEFAULT_ATTRIBUTION),
		# By name *and* URL, for the same reason `instance` is here.
		"styles": dict(STYLES),
		"instance": instance,
		"plain": False,
		**prefer,
	}


@frappe.whitelist(methods=["POST"])
def set_basemap(style: str = "", labels=None, detail: str = "") -> dict:
	"""Change how this workspace's maps are drawn, for everybody on it.

	Guarded on the workspace owner rather than on a doctype permission: this is
	a presentation decision about every screen in the product, which is the same
	thing branding is, and it is reached from a map rather than from the desk.
	"""
	# Imported here rather than at the top: `workspace` reaches into Frappe's
	# timezone tables at import time, and this module is read by tests that
	# stand up neither.
	from oneapp.onespace.workspace import OWNER_ROLE, SUPPORT_ROLE

	if not set(frappe.get_roles()) & {OWNER_ROLE, SUPPORT_ROLE}:
		frappe.throw(_("Only a workspace admin can change how the map is drawn."),
		             frappe.PermissionError)

	row = frappe.get_doc("OneSpace Map Settings")
	if style:
		if style not in STYLES and style not in ("Follow the instance", "Plain"):
			frappe.throw(_("There is no such basemap."))
		row.map_style = style
	if labels is not None:
		row.map_labels = 1 if frappe.parse_json(labels) else 0
	if detail:
		if detail not in ("Full", "Quiet", "Minimal"):
			frappe.throw(_("There is no such level of detail."))
		row.map_detail = detail
	row.save(ignore_permissions=True)
	frappe.clear_document_cache("OneSpace Map Settings")
	return chosen()


def hosts() -> tuple[str, ...]:
	"""The tile hosts this instance actually talks to, for a legal clause.

	Read off the *instance's* configuration rather than off `boot()`, because a
	published agreement cannot depend on what one workspace picked this morning
	— every curated style is the same host, so a workspace switching between
	them changes nothing a clause has to say.
	"""
	if frappe.conf.get("oneapp_map_plain"):
		return ()
	# A configured host is the operator's own business to disclose; naming it
	# here would put a deployment detail into a published agreement.
	if (frappe.conf.get("oneapp_map_style") or "").strip():
		return ()
	if (frappe.conf.get("oneapp_map_tiles") or "").strip():
		return ()
	return DEFAULT_HOSTS

File: frontend/src/modules/onespace/lib/screen/basemap.js
/**
 * The ground a map is drawn on.
 *
 * Two surfaces draw maps — the engine's map view type and OneMobility's live
 * network — and both need the same three answers: what is under the records,
 * who to credit for it, and what to do when there is no tile host at all. One
 * module rather than two copies, because the day an instance is pointed at its
 * own tile store is the day both have to agree.
 *
 * Where the answer comes from: `onespace/basemap.py`, on the boot payload.
 * Which tile store a bench uses is a deployment fact, not a workspace's choice.
 *
 * **Vector, and that is what makes the map customisable at all.** A raster tile
 * is a picture somebody else already drew: the only thing a reader can change
 * about it is what goes on top. A vector style is JSON the browser executes, so
 * whether places are named and how much of the world is drawn under the records
 * are properties on layers — `restyle` below sets them on the running map, with
 * no reload and no second tile store.
 *
 * The old note, kept because the trade it describes is still real:
 * raster and not a vector style URL, by default. A raster source is a URL
 * inside a style we own, so the style always parses and the map always fires
 * `load`; a style *URL* that 404s leaves MapLibre with no style, no `load`
 * event and a screen awaiting a promise that will never settle. A tile that
 * fails is one grey square. Given `style` in site config the operator has
 * chosen the other trade deliberately, and gets it.
 */

import { basemap as configured } from '@/shared/lib/runtime/boot'
import { paintable, tokenInk } from '@/modules/onespace/lib/screen/ink'

/** Is the board dark? Read off a token rather than a media query, because the
 *  workspace's own theme can say dark on a light system and the tiles should
 *  follow what is actually painted. */
export function isDark() {
  const ground = paintable(tokenInk('--surface-base', '#ffffff'), '#ffffff')
  const hex = ground.startsWith('#') ? ground.slice(1) : ''
  if (hex.length !== 6) return false
  const [r, g, b] = [0, 2, 4].map((i) => parseInt(hex.slice(i, i + 2), 16))
  // Rec. 709 luma. Anything below the midpoint is a dark surface.
  return 0.2126 * r + 0.7152 * g + 0.0722 * b < 128
}

/**
 * What this workspace has said about its ground: whether to name places, and
 * how much to draw. Read from the boot payload, so a surface that has no picker
 * still honours the choice made on one that has.
 */
export function preferred() {
  return {
    pick: configured?.pick || 'Follow the instance',
    labels: configured?.labels !== false,
    detail: configured?.detail || 'Quiet',
  }
}

/**
 * Who to credit, for the attribution control — and usually nobody, now.
 *
 * A vector style names a TileJSON, and that document carries the full linked
 * attribution the licence wants. MapLibre reads it and draws it, so anything we
 * add sits *beside* it: the first version of this stacked two credits into one
 * line that said the same thing twice, once less completely. So the server
 * sends one only where there is no document to read it out of — a raster
 * template — or where an operator has named their own.
 */
export function attribution() {
  return (configured?.attribution || '').trim()
}

/** A style with nothing in it but a colour. Ours, so it always parses. */
export function flatGround(background) {
  return {
    version: 8,
    sources: {},
    layers: [{ id: 'ground', type: 'background', paint: { 'background-color': background } }],
  }
}

/**
 * A MapLibre style: the configured one, or ours over the configured tiles, or
 * a flat ground when there are none.
 *
 * @param {string} background a paintable colour for the ground
 * @param {string} override a style URL that wins over everything, or one of two
 *   words. An empty string is "I have no opinion", so a caller that has decided
 *   something cannot say it that way: `'none'` is a flat ground on purpose, and
 *   `'instance'` is what the *instance* configured, ignoring the workspace pick
 *   that `style` on the payload already has baked into it. A picker needs both,
 *   because after it has been used the payload describes the morning.
 */
export function styleFor(background, override = '') {
  if (override === 'none') return flatGround(background)
  const url = String(
    override === 'instance' ? configured?.instance || '' : override || configured?.style || '',
  ).trim()
  if (url) return url

  const tiles = String((isDark() ? configured?.dark : configured?.tiles) || '').trim()
  const style = flatGround(background)
  if (!tiles) return style

  style.sources.basemap = {
    type: 'raster',
    tiles: [tiles],
    tileSize: 256,
    attribution: attribution(),
  }
  // Under everything: `addLayer` without a `before` appends, so every layer a
  // surface adds afterwards lands on top of this without having to say so.
  style.layers.push({ id: 'basemap', type: 'raster', source: 'basemap' })
  return style
}

/**
 * Resolve when the map is ready, or when it plainly is not.
 *
 * `map.on('load')` fires once the style and its first tiles are in. It does
 * not fire at all if the style could not be fetched — a configured style URL
 * behind a firewall, an offline bench — and a screen awaiting it then shows
 * nothing for ever with no error anywhere. Worse, `addSource` on a map with no
 * style throws, so simply carrying on is not enough.
 *
 * So: wait, and if the style never arrived, put the flat ground on instead and
 * wait for that. Whatever else is wrong, the records get drawn.
 */
export function whenLoaded(map, background, within = 8000) {
  const once = (span) =>
    new Promise((resolve) => {
      let done = false
      const settle = () => {
        if (done) return
        done = true
        resolve()
      }
      if (map.isStyleLoaded?.()) return settle()
      map.on('load', settle)
      map.on('idle', settle)
      setTimeout(settle, span)
    })

  return once(within).then(() => {
    if (map.isStyleLoaded?.()) return undefined
    map.setStyle(flatGround(background))
    return once(3000)
  })
}

/** Every host the ground could come from, for `quietTiles`. */
function grounds() {
  const froms = [
    configured?.style,
    configured?.instance,
    configured?.tiles,
    configured?.dark,
    ...Object.values(configured?.styles || {}),
  ]
  const hosts = new Set()
  for (const one of froms) {
    // Absolute only, deliberately. A relative one is a ground we serve
    // ourselves, and a 404 on that is our own bug rather than somebody else's
    // afternoon — it should print.
    try {
      if (one) hosts.add(new URL(one).host)
    } catch {
      // Not an absolute URL. Nothing to match against, which is the safe
      // direction: an error nobody recognised is one that still gets printed.
    }
  }
  return hosts
}

/**
 * Stop a basemap that cannot be reached from shouting about it.
 *
 * MapLibre reports every failed tile to the console as an error, one per tile,
 * a fresh dozen on every pan. On a workspace behind a firewall, on a bench with
 * no route out, or on the day the tile host is having a bad afternoon, that is
 * a console full of red for something the map already handles: the ground is
 * blank and every record is still drawn on it.
 *
 * **Matched on the host, not only on the source id.** With a raster ground the
 * failure is a tile in a source we named `basemap`, so the id was enough. A
 * vector style fails one step earlier — the style *document* cannot be fetched,
 * before there is a source to blame — and that error carries no `sourceId` at
 * all, so switching to vector quietly put a red line under every map screen on
 * every bench with no route out.
 *
 * Only the ground is quietened. An error about a source or a layer we added is
 * ours and still surfaces, because that one means the screen is wrong.
 */
export function quietTiles(map) {
  const hosts = grounds()
  map.on('error', (event) => {
    const source = event?.sourceId
    if (source === 'basemap' || event?.error?.status === 404) return
    const url = event?.error?.url
    if (url) {
      try {
        if (hosts.has(new URL(url).host)) return
      } catch {
        // An error about something that is not a URL is not a ground failure.
      }
    }
    console.error(event?.error || event)
  })
}


/**
 * How much of the world a workspace wants under its records.
 *
 * Matched on layer id prefixes, which is safe in a way it looks like it is not:
 * every style here is built by planetiler against the OpenMapTiles schema, and
 * that schema names its layers. A style that does not follow it simply matches
 * nothing and draws in full, which is the right way for this to fail.
 */
const DECORATION = ['building', 'landcover', 'landuse', 'park', 'poi', 'housenumber', 'aeroway']

/** What a minimal ground keeps. Everything a route needs to be placed, and no more. */
const ESSENTIAL = ['background', 'water', 'waterway', 'road', 'bridge', 'tunnel', 'boundary']

function starts(id, prefixes) {
  const name = String(id || '').toLowerCase()
  return prefixes.some((one) => name.startsWith(one))
}

/**
 * Apply a workspace's basemap preferences to a map that has already loaded.
 *
 * On the running style rather than by fetching and patching the JSON first,
 * which matters twice: it is instant, so a switch in the picker is a switch on
 * the screen, and it cannot fail differently from the style that is actually
 * drawn — there is only one copy.
 *
 * Only the basemap's own layers are touched. Ours are added after the style
 * loads and are passed over by name, because a legend that hides itself when
 * somebody turns off place names would be a surprising way to learn what this
 * setting does.
 */
export function restyle(map, prefer = {}, ours = []) {
  if (!map || !map.getStyle) return
  const style = map.getStyle()
  if (!style?.layers) return

  const labels = prefer.labels !== false
  const detail = prefer.detail || 'Full'
  const mine = new Set(ours)

  for (const layer of style.layers) {
    if (mine.has(layer.id)) continue
    let show = true
    // A label is a symbol layer, in every style built this way.
    if (!labels && layer.type === 'symbol') show = false
    if (detail === 'Quiet' && starts(layer.id, DECORATION)) show = false
    if (detail === 'Minimal' && !starts(layer.id, ESSENTIAL)) show = false
    try {
      map.setLayoutProperty(layer.id, 'visibility', show ? 'visible' : 'none')
    } catch {
      // A layer that has gone while we were walking the list. The next call
      // rebuilds from the style as it is then.
    }
  }
}
